mautrix_signal/portal.py

```diff
@@ -725,6 +725,43 @@ async def handle_matrix_leave(self, user: u.User) -> None:
                 await self.signal.leave_group(user.username, self.chat_id)
             # TODO cleanup if empty
 
+    async def kick_matrix(self, user: u.User | p.Puppet, source: u.User) -> None:
+        try:
+            await self.signal.update_group(
+                source.username, self.chat_id, remove_members=[user.address]
+            )
+        except Exception as e:
+            self.log.exception(f"Failed to kick Signal user: {e}")
+            await self.main_intent.invite_user(
+                self.mxid, user.mxid, check_cache=True, reason=f"Failed to kick Signal user: {e}"
+            )
+            await user.intent_for(self).ensure_joined(self.mxid)
+
+    async def ban_matrix(self, user: u.User | p.Puppet, source: u.User) -> None:
+        try:
+            await self.signal.ban_user(source.username, self.chat_id, users=[user.address])
+        except Exception as e:
+            self.log.exception(f"Failed to ban Signal user: {e}")
+            await self.main_intent.unban_user(self.mxid, user.mxid)
+            info = await self.signal.get_group(source.username, self.chat_id)
+            if user.address in info.members:
+                await self.main_intent.invite_user(
+                    self.mxid,
+                    user.mxid,
+                    check_cache=True,
+                    reason=f"Failed to ban Signal user: {e}",
```

Comment:
I put the reason here because unbans don't come with reasons. If the user wasn't in the room at the time of the failed ban attempt, the error will not be printed to the room. I don't think it's important enough to have the bridge bot send a notification in that case.

Reply:
Unbans can have reasons (might need to update mautrix-python) so it should just be put there

Reply:
It doesn't look like it:
https://github.com/mautrix/python/blob/4ba953d7180bfc0b2d8b822c0701111796b9d6df/mautrix/client/store_updater.py#L113
https://matrix.org/docs/spec/client_server/r0.6.1.html#post-matrix-client-r0-rooms-roomid-unban

Reply:
https://spec.matrix.org/v1.2/client-server-api/#post_matrixclientv3roomsroomidunban

Reply:
apparently I checked the wrong version. I changed it, but I currently can't see the reasons in my client. I'm not sure where the problem lies (most likely in my mautrix-python changes), but I don't think it's here.

Reply:
Correction, it works with the pull request for mautrix-python. Would I have to adjust the version for mautrix-python somewhere in the requirements?

Reply:
works